cpp/CMakeLists.txt

```diff
@@ -330,6 +330,10 @@ if(ARROW_BUILD_BENCHMARKS
   set(ARROW_TESTING ON)
 endif()
 
+if(ARROW_GANDIVA)
+  set(ARROW_WITH_RE2 ON)
+endif()
```

Comment:
How about forcing to enable/disable `ARROW_WITH_RE2` in `ThridpartyToolchain.cmake` like `ARROW_WITH_THRIFT`?